pageserver/src/branches.rs

```diff
@@ -56,39 +58,69 @@ impl BranchInfo {
             })?
             .parse::<ZTimelineId>()?;
 
-        let timeline = match repo.get_timeline(timeline_id)? {
+        let timeline = match repo
+            .get_timeline(timeline_id)
+            .ok_or(anyhow::anyhow!("unknown timeline {}", timeline_id))?
+        {
             RepositoryTimeline::Local(local_entry) => local_entry,
             RepositoryTimeline::Remote { .. } => {
                 bail!("Timeline {} is remote, no branches to display", timeline_id)
             }
         };
 
-        // we use ancestor lsn zero if we don't have an ancestor, so turn this into an option based on timeline id
-        let (ancestor_id, ancestor_lsn) = match timeline.get_ancestor_timeline_id() {
-            Some(ancestor_id) => (
-                Some(ancestor_id.to_string()),
-                Some(timeline.get_ancestor_lsn().to_string()),
-            ),
-            None => (None, None),
+        let branch_info = match timeline {
+            LocalRepositoryTimeline::Loaded(timeline) => {
+                // non incremental size calculation can be heavy, so let it be optional
+                // needed for tests to check size calculation
+                let current_logical_size_non_incremental = include_non_incremental_logical_size
+                    .then(|| {
+                        timeline.get_current_logical_size_non_incremental(
+                            timeline.get_last_record_lsn(),
+                        )
+                    })
+                    .transpose()?;
+
+                // we use ancestor lsn zero if we don't have an ancestor, so turn this into an option based on timeline id
+                let (ancestor_id, ancestor_lsn) = match timeline.get_ancestor_timeline_id() {
+                    Some(ancestor_id) => (
+                        Some(ancestor_id.to_string()),
+                        Some(timeline.get_ancestor_lsn().to_string()),
+                    ),
+                    None => (None, None),
+                };
+
+                BranchInfo {
+                    name,
+                    timeline_id,
+                    latest_valid_lsn: timeline.get_last_record_lsn(),
+                    ancestor_id,
+                    ancestor_lsn,
+                    current_logical_size: Some(timeline.get_current_logical_size()),
+                    current_logical_size_non_incremental,
+                }
+            }
+            LocalRepositoryTimeline::Unloaded { metadata } => {
+                // TODO is there some generic enough way to avoid this duplication? generic function with common trait?
```

Comment:
Good Idea! It would also help to avoid error checking noise around ancestor fields, because ancestor id is Option, but ancestor lsn isnt and is set to zero when ancestor_id is None